File: Device/Luckfoxpico86/code/python_engine/src/xsolar_bridge.py
"""
Xsolar Bridge - Giao tiếp với xsolar cloud qua MQTT

Chức năng:
1. Periodic push device states lên xsolar mỗi 10 phút
2. Event-driven push khi device state thay đổi
3. Nhận lệnh điều khiển từ xsolar (smarteos/bluCafe/+/set)
4. Convert lệnh xsolar thành ZbSend và gửi về local broker
"""

import logging
import json
from datetime import datetime
from typing import Dict, Any, Optional

logger = logging.getLogger(__name__)


class XsolarBridge:
    """Bridge giữa BMS Engine và xsolar cloud"""
    
    def __init__(self, mqtt_local, mqtt_xsolar, state_manager, device_manager):
        """
        Khởi tạo Xsolar Bridge
        
        Args:
            mqtt_local: MQTT client cho local broker (NanoMQ)
            mqtt_xsolar: MQTT client cho xsolar broker
            state_manager: State manager instance
            device_manager: Device manager instance
        """
        self.mqtt_local = mqtt_local
        self.mqtt_xsolar = mqtt_xsolar
        self.state_manager = state_manager
        self.device_manager = device_manager
        
        # Track last push time per device để tránh spam
        self._last_push: Dict[str, float] = {}
        self._min_push_interval = 5  # Tối thiểu 5 giây giữa 2 lần push cùng device
        
        logger.info("Xsolar Bridge initialized")
    
    def start(self):
        """
        Start Xsolar Bridge - subscribe nhận lệnh từ xsolar
        """
        # Subscribe remote commands từ xsolar
        # Format: smarteos/bluCafe/{device_id}/set
        # Payload: {"power": "ON", "temperature": 24, "fan": 2}
        self.mqtt_xsolar.subscribe("smarteos/bluCafe/+/set", qos=1)
        
        # Set message handler cho xsolar client
        self.mqtt_xsolar.set_message_handler(self._handle_xsolar_message)
        
        logger.info("Xsolar Bridge started - subscribed to smarteos/bluCafe/+/set")
    
    def _handle_xsolar_message(self, topic: str, payload: Any):
        """
        Xử lý message từ xsolar broker
        
        Args:
            topic: MQTT topic
            payload: Message payload
        """
        try:
            parts = topic.split('/')
            
            # Handle remote commands: smarteos/bluCafe/{device_id}/set
            if len(parts) == 4 and parts[0] == 'smarteos' and parts[1] == 'bluCafe' and parts[3] == 'set':
                device_id = parts[2]
                self._handle_remote_command(device_id, payload)
            
        except Exception as e:
            logger.error(f"Error handling xsolar message: {e}", exc_info=True)
    
    def _handle_remote_command(self, device_id: str, payload: Any):
        """
        Xử lý lệnh điều khiển từ xsolar
        
        Args:
            device_id: Zigbee address (VD: "0xC5A9")
            payload: Command payload (dict hoặc JSON string)
        """
        # Parse payload
        if isinstance(payload, str):
            try:
                cmd = json.loads(payload)
            except json.JSONDecodeError:
                logger.warning(f"Invalid JSON from xsolar: {payload}")
                return
        else:
            cmd = payload
        
        # Lookup device
        device = self.device_manager.get_device(device_id)
        if not device:
            logger.warning(f"Remote command: device not found: {device_id}")
            return
        
        if device['nr_type'] != 'ac_controller':
            logger.warning(f"Remote command: only AC supported, got {device['nr_type']}")
            return
        
        gateway = device['gateway']
        attributes = device['attributes']
        
        # Build ZbSend writes
        writes = {}
        
        # Power
        if 'power' in cmd:
            power_attr_id = self._find_attr_id_by_xsolar_key(attributes, 'power')
            if power_attr_id:
                value = 1 if cmd['power'] in ('ON', True, 1) else 0
                writes[f"EF00/{power_attr_id}"] = value
        
        # Temperature
        if 'temperature' in cmd:
            temp_attr_id = self._find_attr_id_by_xsolar_key(attributes, 'temperature')
            if temp_attr_id:
                writes[f"EF00/{temp_attr_id}"] = int(cmd['temperature'])
        
        # Fan speed
        if 'fan' in cmd:
            fan_attr_id = self._find_attr_id_by_xsolar_key(attributes, 'fan')
            if fan_attr_id:
                writes[f"EF00/{fan_attr_id}"] = int(cmd['fan'])
        
        if not writes:
            logger.warning(f"Remote command: no valid commands for {device_id}")
            return
        
        # Publish ZbSend về local broker
        zb_send_payload = {
            "Device": device_id,
            "Write": writes
        }
        
        topic = f"cmnd/{gateway}/ZbSend"
        self.mqtt_local.publish(topic, json.dumps(zb_send_payload), qos=1)
        
        # Log command
        self.state_manager.log_event(
            device_id,
            'remote_command',
            {'source': 'xsolar', 'command': cmd, 'writes': writes},
            device['nr_type'],
            device['group']
        )
        
        logger.info(f"Remote command from xsolar: {device_id} -> {cmd}")
    
    def push_device_state(self, device_addr: str):
        """
        Push device state lên xsolar (event-driven)
        
        Args:
            device_addr: Zigbee address
        """
        device = self.device_manager.get_device(device_addr)
        if not device:
            return
        
        # Rate limiting - tránh spam
        now = datetime.now().timestamp()
        last_push = self._last_push.get(device_addr, 0)
        if now - last_push < self._min_push_interval:
            return
        
        state = self.state_manager.get_latest_state(device_addr)
        if not state:
            return
        
        # Build payload với chỉ xsolar-marked attributes
        xsolar_data = self._build_xsolar_data(device, state)
        if not xsolar_data:
            return
        
        payload = {
            'ts': datetime.now().isoformat() + '+07:00',
            'site': 'bluCafe',
            'id': device_addr,
            'type': device['nr_type'],
            'name': device['name'],
            'location': device['group'],
            'data': xsolar_data
        }
        
        topic = f"smarteos/bluCafe/{device_addr}"
        
        try:
            self.mqtt_xsolar.publish(topic, json.dumps(payload), qos=0)
            self._last_push[device_addr] = now
            logger.info(f"Event push: {device_addr} -> xsolar ({len(xsolar_data)} attrs)")
        except Exception as e:
            logger.error(f"Failed to push {device_addr} to xsolar: {e}")
    
    def push_all_states(self):
        """
        Push tất cả device states lên xsolar (periodic - gọi bởi scheduler)
        """
        logger.info("Periodic push: pushing all device states to xsolar...")
        
        for device_addr, device in self.device_manager.get_all_devices().items():
            state = self.state_manager.get_latest_state(device_addr)
            if not state:
                logger.debug(f"No state for {device_addr}, skipping")
                continue
            
            xsolar_data = self._build_xsolar_data(device, state)
            if not xsolar_data:
                logger.debug(f"No xsolar data for {device_addr}, skipping")
                continue
            
            payload = {
                'ts': datetime.now().isoformat() + '+07:00',
                'site': 'bluCafe',
                'id': device_addr,
                'type': device['nr_type'],
                'name': device['name'],
                'location': device['group'],
                'data': xsolar_data
            }
            
            topic = f"smarteos/bluCafe/{device_addr}"
            
            try:
                self.mqtt_xsolar.publish(topic, json.dumps(payload), qos=0)
                logger.info(f"Periodic push: {device_addr} -> xsolar ({len(xsolar_data)} attrs)")
            except Exception as e:
                logger.error(f"Failed to push {device_addr} to xsolar: {e}")
        
        self._last_push = {}  # Reset rate limit sau periodic push
    
    def _build_xsolar_data(self, device: Dict, state: Dict) -> Dict:
        """
        Build xsolar data dict từ device state
        
        Chỉ lấy attributes có xsolar: true, dùng xsolar_key làm key name.
        
        Args:
            device: Device dictionary từ device_manager
            state: State dictionary từ state_manager
            
        Returns:
            Dictionary với xsolar_key -> value
        """
        xsolar_data = {}
        
        for attr_id, attr_config in device['attributes'].items():
            if not attr_config.get('xsolar', False):
                continue
            
            # Handle composite attributes (MCB dp6)
            if attr_config.get('decode'):
                for rule in attr_config['decode']:
                    if rule.get('xsolar', False) and rule['id'] in state:
                        xsolar_data[rule['xsolar_key']] = state[rule['id']]
            elif attr_config['label'] in state:
                xsolar_key = attr_config.get('xsolar_key')
                if xsolar_key:
                    xsolar_data[xsolar_key] = state[attr_config['label']]
        
        return xsolar_data
    
    def _find_attr_id_by_xsolar_key(self, attributes: Dict, xsolar_key: str) -> Optional[str]:
        """
        Tìm attribute ID theo xsolar_key
        
        Args:
            attributes: Attributes dictionary từ device
            xsolar_key: xsolar key name (power, temperature, fan)
            
        Returns:
            Attribute ID hoặc None
        """
        for attr_id, attr_config in attributes.items():
            if attr_config.get('xsolar_key') == xsolar_key:
                return attr_id
        
        return None

File: Device/Luckfoxpico86/code/Rush_engine/src/xsolar_bridge.rs
use dashmap::DashMap;
use log::{debug, info, warn};
use rumqttc::QoS;
use serde_json::Value;
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::Mutex;

use crate::device_manager::SharedDeviceManager;
use crate::mqtt_client::MqttClient;
use crate::state_manager::StateManager;

pub struct XsolarBridge {
    mqtt_local: Arc<MqttClient>,
    mqtt_xsolar: Arc<MqttClient>,
    state_manager: Arc<Mutex<StateManager>>,
    device_manager: SharedDeviceManager,
    last_push: Mutex<HashMap<String, f64>>,
    min_push_interval: f64,
    /// In-memory cache of the last known state per device (zigbee addr ->
    /// label -> value). Written by the ingest fan-out, read by both the
    /// real-time push and the periodic 10-min snapshot push. Never reads DB.
    cache: DashMap<String, HashMap<String, Value>>,
}

impl XsolarBridge {
    pub fn new(
        mqtt_local: Arc<MqttClient>,
        mqtt_xsolar: Arc<MqttClient>,
        state_manager: Arc<Mutex<StateManager>>,
        device_manager: SharedDeviceManager,
    ) -> Self {
        Self {
            mqtt_local,
            mqtt_xsolar,
            state_manager,
            device_manager,
            last_push: Mutex::new(HashMap::new()),
            min_push_interval: 5.0,
            cache: DashMap::new(),
        }
    }

    pub async fn start(&self) {
        self.mqtt_xsolar
            .subscribe("smarteos/bluCafe/+/set", QoS::AtLeastOnce)
            .await;
        info!("Xsolar Bridge started - subscribed to smarteos/bluCafe/+/set");
    }

    pub async fn handle_xsolar_message(&self, topic: &str, payload: &Value) {
        let parts: Vec<&str> = topic.split('/').collect();

        if parts.len() == 4
            && parts[0] == "smarteos"
            && parts[1] == "bluCafe"
            && parts[3] == "set"
        {
            let device_id = parts[2];
            self.handle_remote_command(device_id, payload).await;
        }
    }

    async fn handle_remote_command(&self, device_id: &str, payload: &Value) {
        let cmd: Value = match payload {
            Value::String(s) => {
                match serde_json::from_str::<Value>(s) {
                    Ok(v) => v,
                    Err(_) => {
                        warn!("Invalid JSON from xsolar: {}", s);
                        return;
                    }
                }
            }
            other => other.clone(),
        };

        let dm = self.device_manager.read().await;
        let device = match dm.get_device(device_id) {
            Some(d) => d,
            None => {
                warn!("Remote command: device not found: {}", device_id);
                return;
            }
        };

        let mut writes = HashMap::new();

        for (attr_id, attr_config) in &device.attributes {
            let label = &attr_config.label;
            if let Some(raw_val) = cmd.get(label) {
                let value = if attr_config.attr_type == "bool" {
                    match raw_val {
                        Value::String(s) if s.to_uppercase() == "ON" || s == "TRUE" || s == "1" => serde_json::json!(1),
                        _ => serde_json::json!(0),
                    }
                } else {
                    match raw_val {
                        Value::Number(n) => serde_json::json!(n.as_f64().unwrap_or(0.0) as i64),
                        Value::String(s) => serde_json::json!(s.parse::<f64>().unwrap_or(0.0) as i64),
                        _ => serde_json::json!(0),
                    }
                };
                writes.insert(format!("EF00/{}", attr_id), value);
            }

            if let Some(decode_rules) = &attr_config.decode {
                for rule in decode_rules {
                    let rlabel = rule.label.as_deref().unwrap_or("");
                    if let Some(raw_val) = cmd.get(rlabel) {
                        let value = match raw_val {
                            Value::Number(n) => serde_json::json!(n.as_f64().unwrap_or(0.0) as i64),
                            Value::String(s) => serde_json::json!(s.parse::<f64>().unwrap_or(0.0) as i64),
                            _ => serde_json::json!(0),
                        };
                        writes.insert(format!("EF00/{}", attr_id), value);
                    }
                }
            }
        }

        if writes.is_empty() {
            warn!("Remote command: no matching attributes for {}", device_id);
            return;
        }

        let zb_send = serde_json::json!({
            "Device": device_id,
            "Write": writes,
            "Endpoint": 1,
        });

        let topic = format!("cmnd/{}/ZbSend", device.gateway);
        self.mqtt_local
            .publish_json(&topic, &zb_send, QoS::AtLeastOnce, false)
            .await;

        let sm = self.state_manager.lock().await;
        sm.log_event(
            device_id,
            "remote_command",
            &serde_json::json!({"source": "xsolar", "command": cmd, "writes": writes}),
            Some(&device.nr_type),
            Some(&device.group),
        );

        info!(
            "Remote command from xsolar: {} {} -> {:?}",
            device_id, device.name, cmd
        );
    }

    /// Write-only cache update from the ingest fan-out (no network, no backpressure).
    pub fn update_cache(&self, device_addr: &str, updates: &HashMap<String, Value>) {
        let mut entry = self.cache.entry(device_addr.to_string()).or_default();
        for (k, v) in updates {
            entry.insert(k.clone(), v.clone());
        }
    }

    /// Seed the cache from SQLite once at engine startup so devices that have
    /// not yet reported after a restart still appear in the periodic push
    /// (review §4). Uses the fixed window-function query via get_latest_state().
    pub async fn seed_from_db(&self) {
        let addrs: Vec<String> = {
            let dm = self.device_manager.read().await;
            dm.get_all_devices().keys().cloned().collect()
        };
        let sm = self.state_manager.lock().await;
        for addr in addrs {
            let state = sm.get_latest_state(&addr);
            if !state.is_empty() {
                self.update_cache(&addr, &state);
            }
        }
        info!("Xsolar Bridge: seeded cache for {} devices", self.cache.len());
    }

    /// Real-time on-change push, throttled per device. Reads the in-memory cache.
    pub async fn push_device_state(&self, device_addr: &str) {
        let dm = self.device_manager.read().await;
        let device = match dm.get_device(device_addr) {
            Some(d) => d,
            None => return,
        };

        let now = chrono::Utc::now().timestamp() as f64;
        {
            let last_push = self.last_push.lock().await;
            let last = last_push.get(device_addr).copied().unwrap_or(0.0);
            if now - last < self.min_push_interval {
                return;
            }
        }

        let state = self.cache.get(device_addr).map(|r| r.clone()).unwrap_or_default();
        if state.is_empty() {
            return;
        }

        let xsolar_data = build_xsolar_data(device, &state);
        if xsolar_data.is_empty() {
            return;
        }

        let payload = serde_json::json!({
            "ts": format!("{}+07:00", chrono::Local::now().format("%Y-%m-%dT%H:%M:%S")),
            "site": "bluCafe",
            "id": device_addr,
            "type": device.nr_type,
            "name": device.name,
            "location": device.group,
            "data": xsolar_data,
        });

        let topic = format!("smarteos/bluCafe/{}", device_addr);
        self.mqtt_xsolar
            .publish_json(&topic, &payload, QoS::AtMostOnce, false)
            .await;

        {
            let mut last_push = self.last_push.lock().await;
            last_push.insert(device_addr.to_string(), now);
        }

        info!(
            "Event push: {} -> xsolar ({} attrs)",
            device_addr,
            xsolar_data.len()
        );
    }

    /// Periodic full-snapshot push (default every 10 min). Reads the in-memory
    /// cache, one publish per device — no DB reads on the hot path.
    pub async fn push_all_states(&self) {
        info!("Periodic push: pushing all device states to xsolar...");
        let devices: Vec<crate::device_manager::Device> = {
            let dm = self.device_manager.read().await;
            dm.get_all_devices().values().cloned().collect()
        };

        for device in &devices {
            let device_addr = &device.zigbee_addr;
            let state = self.cache.get(device_addr).map(|r| r.clone()).unwrap_or_default();
            if state.is_empty() {
                debug!("No state for {}, skipping", device_addr);
                continue;
            }

            let xsolar_data = build_xsolar_data(device, &state);
            if xsolar_data.is_empty() {
                debug!("No xsolar data for {}, skipping", device_addr);
                continue;
            }

            let payload = serde_json::json!({
                "ts": format!("{}+07:00", chrono::Local::now().format("%Y-%m-%dT%H:%M:%S")),
                "site": "bluCafe",
                "id": device_addr,
                "type": device.nr_type,
                "name": device.name,
                "location": device.group,
                "data": xsolar_data,
            });

            let topic = format!("smarteos/bluCafe/{}", device_addr);
            let _ = self
                .mqtt_xsolar
                .publish_json(&topic, &payload, QoS::AtMostOnce, false)
                .await;

            info!(
                "Periodic push: {} -> xsolar ({} attrs)",
                device_addr,
                xsolar_data.len()
            );
        }

        self.last_push.lock().await.clear();
    }
}

fn build_xsolar_data(
    device: &crate::device_manager::Device,
    state: &HashMap<String, Value>,
) -> serde_json::Map<String, Value> {
    let mut data = serde_json::Map::new();

    for (_attr_id, attr_config) in &device.attributes {
        if !attr_config.xsolar {
            continue;
        }

        if let Some(decode_rules) = &attr_config.decode {
            for rule in decode_rules {
                if rule.xsolar {
                    if let Some(value) = state.get(&rule.id) {
                        data.insert(
                            rule.label.clone().unwrap_or_else(|| rule.id.clone()),
                            value.clone(),
                        );
                    }
                }
            }
        } else if let Some(value) = state.get(&attr_config.label) {
            data.insert(attr_config.label.clone(), value.clone());
        }
    }

    data
}
